README.md
# Cephas

A Python Mongo REST Service based on a schema to manage 'where, when, and what' information.

The idea is to be able to track events in time, where they happen(ed) and what the details were.
For example:
- Used to manage historical data
- Used to manage future events

# Development

Example running locally (Windows):
```
python3 -m venv .venv
.\.venv\Scripts\Activate.ps1
pip install -r .\requirements.txt
python3 main.py
```

# Testing

A [Robot Framework](https://robotframework.org/) acceptance test can be run as follows (from the virtual env):
```
pip install -r .\test-requirements.txt
cd tests
robot .\basic_crud.robot
```

# Containerization

Build the image `docker build -t cephas:latest .`

Run the image `docker run -p 8080:8080 cephas:latest`

Run with compose `docker-compose up`

## Container Environment

- **CEPHAS_SERVER_LOGLEVEL**: Server logging output level (default DEBUG)
- **CEPHAS_SERVER_LISTEN_HOST**: Hostname or IP where the service will bind too
- **CEPHAS_SERVER_LISTEN_PORT**: Port number where the service will listen for requests
- **CEPHAS_DB_NAME**: Name of the MongoDB database
- **CEPHAS_DB_COLLECTION_NAME**: Name of the MongoDB collection
- **CEPHAS_DB_HOST**: Hostname where the MongoDB service is running
- **CEPHAS_DB_PORT**: Port number where the MongoDB is listening for connections
- **CEPHAS_DB_USER**: Mongo client username
- **CEPHAS_DB_PASS**: Mongo client password
- **CEPHAS_DB_SOCKET_TIMEOUT_MS**: Mongo client socketTimeoutMS
- **CEPHAS_DB_CONNECT_TIMEOUT_MS**: Mongo client connectTimeoutMS


# When Where What
There are nuances to managing data with respect to time and space.
Following are some notes to be converted into documentation details.

## Time
https://docs.mongodb.com/manual/tutorial/model-time-data/

Stored in UTC...local time can be computed by also storing the event timezone.
A strategy must be developed for managing time relative to location.
There are three possible perspectives of time:
- UTC for standardized records and comparison
- Local time for user presentation
- Time relative to the location where the event is

## Space
The application currently supports point locatoin data.
This should be expanded to fully support the geoJson specification.

https://docs.mongodb.com/manual/tutorial/geospatial-tutorial/
http://tugdualgrall.blogspot.com/2014/08/introduction-to-mongodb-geospatial.html

Note: coordinates are  [longitude, latitude]  ... google maps is  [latitude, longitude] apparently
https://stackoverflow.com/questions/41513112/cant-extract-geo-keys-longitude-latitude-is-out-of-bounds


# Example in Mongo

A day at Disneyland

## Create

```
db.www.insert(
    {
        "key": "abc123",
        "deleted": false,
        "begin_timestamp": ISODate("2020-11-10T09:00:00.000Z"),
        "end_timestamp": ISODate("2020-11-10T09:30:00.000Z"),
        "location": {
            "type": "Point",
            "coordinates": [-117.923667, 33.809173]
        },
        "description": "Family vacation to Disneyland. Started the morning in the Downtown Disney District."
    }
)

db.www.insert(
    {
        "key": "def234",
        "deleted": false,
        "begin_timestamp": ISODate("2020-11-10T09:30:00.000Z"),
        "end_timestamp": ISODate("2020-11-10T10:00:00.000Z"),
        "location": {
            "type": "Point",
            "coordinates": [-117.920500, 33.811265]
        },
        "description": "Family vacation to Disneyland. Went to Tarzan's Treehouse."
    }
)

db.www.insert(
    {
        "key": "ghi345",
        "deleted": false,
        "begin_timestamp": ISODate("2020-11-10T10:00:00.000Z"),
        "end_timestamp": ISODate("2020-11-10T10:30:00.000Z"),
        "location": {
            "type": "Point",
            "coordinates": [-117.920814, 33.811299]
        },
        "description": "Family vacation to Disneyland. Time for Pirates of the Caribbean. Arrr!"
    }
)

db.www.insert(
    {
        "key": "jkl456",
        "deleted": false,
        "begin_timestamp": ISODate("2020-11-10T10:30:00.000Z"),
        "end_timestamp": ISODate("2020-11-10T11:00:00.000Z"),
        "location": {
            "type": "Point",
            "coordinates": [-117.920814, 33.811299]
        },
        "description": "Family vacation to Disneyland. Isla and I went on Pirates of the Caribbean again."
    }
)

db.www.insert(
    {
        "key": "mno567",
        "deleted": false,
        "begin_timestamp": ISODate("2020-11-10T10:30:00.000Z"),
        "end_timestamp": ISODate("2020-11-10T11:00:00.000Z"),
        "location": {
            "type": "Point",
            "coordinates": [-117.921460, 33.811210]
        },
        "description": "Family vacation to Disneyland. Ellen and Josie went to New Orleans Square to wait for us."
    }
)
```

## Read

`db.www.find({"key": "abc123"}).pretty();`

## Update
The application supports updating of where, when and what data.
A record of 'last_updated' is managed in the persistence layer.

```
db.www.update(
    {"key": "abc123"},
    {
        "$set": {
            "last_updated": new Date(),
            "description": "Family vacation to Disneyland. Started the morning in the Downtown Disney District to get coffee."
        }
    }
)
```

## Delete
No actual deletes, just using tombstoning protocal in the persistence layer.

```
db.www.update(
    {"key": "abc123"},
    {
        "$set": {
            "last_updated": new Date(),
            "deleted": true
        }
    }
)
```


# Indexes

Indexing here is used for record uniqueness, location operations, and full text searching.

`db.www.getIndexes()`

## Unique
The application generates a unique key for each record using UUID.
This hides the actual persistence ID (i.e. Mongo Object ID) from the client.
To guarantee uniquness the key is indexed.

`db.www.ensureIndex({"key" : 1}, {"unique" : true})`

## Time
Time searches will be a critical part of the application.
Research is needed to determine the benefit and trade-off for indexing chronological data.

## Geo
The application currently supports Point data using geoJSON feature of MongoDB.

`db.www.ensureIndex({ "location":  "2dsphere" })`

## Full text

`db.www.ensureIndex({"description" : "text"})`


# Queries

Given: A time
Then: What was happening everywhere?

```
db.www.find({
    "begin_timestamp":{
        $gte: ISODate("2020-11-10T10:00"),
        $lt: ISODate("2020-11-10T11:00") 
    }
}).pretty();
```

Given: A location
Then: What happened in time?

```
db.www.find({ 
    "location": { 
        $nearSphere: { 
            $geometry: {
                type: "Point",
                coordinates: [-117.920814, 33.811299]
            },
            $maxDistance: 10
        } 
    } 
}).pretty();
```

Given: A time and location
Then: What is happening within a given distance?

```
db.www.find({ 
    "begin_timestamp":{
        $gte: ISODate("2020-11-10T10:00"),
        $lt: ISODate("2020-11-10T11:00") 
    },
    "location": { 
        $nearSphere: { 
            $geometry: {
                type: "Point",
                coordinates: [-117.920814, 33.811299]
            },
            $maxDistance: 10
        } 
    } 
}).pretty();
```


CHANGELOG.md
# Change Log

## Planning

**Features and Improvements**

- Implement full text search
- Errors during persist operations should return dictionary with error message, not just error string
- Search with no match should return dictionary with error message, not just error string
- Add verification to ensure be begin_timestamp actually comes before the end_timestamp
- Update documentation

**Additional Testing**

- Invalid methods
- Missing content-type
- Get deleted data
- Error response format
- Empty results format
- Check UUID searches case insensitive


## [0.0.2] - Not yet released

### Added

- Input data validation and filtering to only allow validated data to be persisted.
- Added example Postman collection for testing and documentation purposes.
- Implemented delete 'tombstoning' pattern so documents aren't actually deleted, they are flagged and filtered as deleted.
- Implemented acceptance test cases using [Robot Framework](https://robotframework.org/)
- Containerization support
- Application environment configuration
- Support mongo client authentication
- Configuration for mongo client connect and socket timeout
- Configuration for log level

### Changed

- Refactored database schema from stock data to 'when/where/what' data
- Refactored code into modules
- Modified MongoContext so 'internal' low level methods start with _. (Example: findOne became _findOne)

### Removed

- Removed extraneous persistence methods in MongoContext (such as distinct, deleteMany...)

### Fixed

- Security Bug: Can no longer save 'anything' in the document

### Known Issues


## [0.0.1] - 2020-06-21
Initial implementation.


cephas/domain.py
# Author:  Tim Alberts (timothy.alberts@snhu.edu)
# Class: CS-499-X2150 Computer Science Capstone 20EW2
# Instructor: Joe Conlan (j.conlan@snhu.edu)
# Due Date: 2020-11-29

import datetime
import geojson
import json
import re
import uuid
from dateutil.parser import *


ERROR_MSG_MISSING = "Missing"
ERROR_MSG_INVALID = "Invalid"
DESCRIPTION_MAX_LENGTH = 1024
DESCRIPTION_REGEX = "[A-Za-z0-9 ,.'!?]*"


def generateKey():
    return str(uuid.uuid4())


def validTime(time_string):
    """ Parse and validate an ISO datetime string
    returns a tuple of boolean flag indicating validity, and the datetime
    """
    try:
        datetime_object = isoparse(time_string)
        return True, datetime_object
    except ValueError:
        return False, None


def validPointCoordsList(coords_list):
    """ Parse and validate geoJson point coordinates from a list
    returns a tuple of boolean flag indicating validity, and the geojson Point
    """
    if type(coords_list) is not list:
        return False, None
    if len(coords_list) != 2:
        return False, None
    try:
        geo_point = geojson.Point((
            coords_list[0],
            coords_list[1]
        ))
        if geo_point.is_valid:
            return True, geo_point
        else:
            return False, None
    except:
        return False, None


def validDescription(description_string):
    """ Trim and validate the description content
    returns a tuple of boolean flag indicationg validity, and the description
    """
    stripped_description_string = description_string.strip()
    if len(stripped_description_string) > DESCRIPTION_MAX_LENGTH:
        return False, None
    if re.fullmatch(DESCRIPTION_REGEX, stripped_description_string):
        return True, stripped_description_string
    else:
        return False, None


class SpaceTimeEvent(object):
    def __init__(self):
        self.valid = False
        self.errorsDict = {}
        self.key = None
        self.deleted = False
        self.begin_timestamp = None
        self.end_timestamp = None
        self.location_type = None
        self.location_coords = None
        self.description = None

    def parseCreateJson(self, json_record):
        self.valid = True

        if 'begin_timestamp' not in json_record:
            self.errorsDict['begin_timestamp'] = ERROR_MSG_MISSING
            self.valid = False
        else:
            valid_begin, self.begin_timestamp = validTime(json_record['begin_timestamp'])
            if not valid_begin:
                self.errorsDict['begin_timestamp'] = ERROR_MSG_INVALID
                self.valid = False

        if 'end_timestamp' in json_record:
            valid_end, self.end_timestamp = validTime(json_record['end_timestamp'])
            if not valid_end:
                self.errorsDict['end_timestamp'] = ERROR_MSG_INVALID
                self.valid = False
        
        if 'location' not in json_record:
            self.errorsDict['location'] = ERROR_MSG_MISSING
            self.valid = False
        else:
            if 'type' not in json_record['location']:
                self.errorsDict['location.type'] = ERROR_MSG_MISSING
                self.valid = False
            else:
                if json_record['location']['type'].lower() != 'point':
                    self.errorsDict['location.type'] = ERROR_MSG_INVALID
                    self.valid = False
                else:
                    self.location_type = 'point'
            if 'coordinates' not in json_record['location']:
                self.errorsDict['location.coordinates'] = ERROR_MSG_MISSING
                self.valid = False
            else:
                valid_location, self.location_coords = validPointCoordsList(
                    json_record['location']['coordinates']
                )
                if not valid_location:
                    self.errorsDict['location.coordinates'] = ERROR_MSG_INVALID
                    self.valid = False

        if 'description' not in json_record:
            self.errorsDict['description'] = ERROR_MSG_MISSING
            self.valid = False
        else:
            valid_description, self.description = validDescription(json_record['description'])
            if not valid_description:
                self.errorsDict['description'] = ERROR_MSG_INVALID
                self.valid = False
        
        if self.valid:
            self.key = generateKey()
    

    def parseUpdateJson(self, json_record):
        self.valid = True
        update_field_count = 0

        if 'begin_timestamp' in json_record:
            update_field_count += 1
            valid_begin, self.begin_timestamp = validTime(json_record['begin_timestamp'])
            if not valid_begin:
                self.errorsDict['begin_timestamp'] = ERROR_MSG_INVALID
                self.valid = False
    
        if 'end_timestamp' in json_record:
            update_field_count += 1
            valid_end, self.end_timestamp = validTime(json_record['end_timestamp'])
            if not valid_end:
                self.errorsDict['end_timestamp'] = ERROR_MSG_INVALID
                self.valid = False
        
        if 'location' in json_record:
            update_field_count += 1
            if 'type' not in json_record['location']:
                self.errorsDict['location.type'] = ERROR_MSG_MISSING
                self.valid = False
            else:
                if json_record['location']['type'].lower() != 'point':
                    self.errorsDict['location.type'] = ERROR_MSG_INVALID
                    self.valid = False
                else:
                    self.location_type = 'point'
            if 'coordinates' not in json_record['location']:
                self.errorsDict['location.coordinates'] = ERROR_MSG_MISSING
                self.valid = False
            else:
                valid_location, self.location_coords = validPointCoordsList(
                    json_record['location']['coordinates']
                )
                if not valid_location:
                    self.errorsDict['location.coordinates'] = ERROR_MSG_INVALID
                    self.valid = False

        if 'description' in json_record:
            update_field_count += 1
            valid_description, self.description = validDescription(json_record['description'])
            if not valid_description:
                self.errorsDict['description'] = ERROR_MSG_INVALID
                self.valid = False
        
        # Must have at least one field to update
        if update_field_count == 0:
            self.errorsDict['update_fields'] = "None"
            self.valid = False
        

    def __repr__(self):
        return "SpaceTimeEvent({}, {}, {}, {}, {}, {}, {}, {})".format(
            self.key,
            self.begin_timestamp,
            self.end_timestamp,
            self.location_type,
            self.location_coords,
            self.description,
            self.valid,
            self.errorsDict
        )


if __name__ == '__main__':
    # A couple simple tests scenarios to play with during development.
    jsonString = """{
        "begin_timestamp": "2020-11-10T09:00:00.000Z",
        "end_timestamp": "2020-11-10T09:30:00.000Z",
        "location": {
            "type": "Point",
            "coordinates": [-117.923667, 33.809173]
        },
        "description": "Some thing happened."
    }"""

    create_record = SpaceTimeEvent()
    create_record.parseCreateJson(json.loads(jsonString))
    print(create_record)

    jsonString = """{
        "key": "abc123",
        "begin_timestamp": "2020-11-10T09:00:00.000Z",
        "end_timestamp": "2020-11-10T09:30:00.000Z",
        "location": {
            "type": "Point",
            "coordinates": [-117.923667, 33.809173]
        },
        "description": "Some thing happened."
    }"""

    update_record = SpaceTimeEvent()
    update_record.parseUpdateJson(json.loads(jsonString))
    print(update_record)


cephas/persist.py
# Author:  Tim Alberts (timothy.alberts@snhu.edu)
# Class: CS-499-X2150 Computer Science Capstone 20EW2
# Instructor: Joe Conlan (j.conlan@snhu.edu)
# Due Date: 2020-11-29

import datetime
import logging
from bson import json_util
from pymongo import MongoClient
from urllib.parse import quote_plus
from . import domain


my_mongo_ctx = None

DEFAULT_PROJECTION = {
    "_id": 0,
    "key": 1,
    "begin_timestamp": 1,
    "end_timestamp": 1,
    "location": 1,
    "description": 1,
}


class MongoConfig(object):
    """ Configuration wrapper for a basic MongoDB connection to a single 
    collection.
    """

    def __init__(self, db_name, host='localhost', port=27017, user=None, 
                    passwd=None, socket_timeout=3000, connect_timeout=3000):
        self.db_name = db_name
        self.host = host
        self.port = port
        self.user = user
        self.passwd = passwd
        self.socket_timeout = socket_timeout
        self.connect_timeout = connect_timeout
    
    def uses_auth(self):
        return self.user is not None and self.passwd is not None
        
    def __repr__(self):
        return 'MongoConfig[db_name=%s,host=%s,port=%s,user=%s,passwd=%s,socket_timeout=%s,connect_timeout=%s]' % (
            str(self.db_name),
            str(self.host),
            str(self.port),
            str(self.user),
            str(self.passwd),
            str(self.socket_timeout),
            str(self.connect_timeout)
        )


class MongoContext(object):
    """ Thin wrapper for a basic MongoDB connection.  Handles the persistence
    logic for the SpaceTimeEvent domain and general mongo interactions
    handling.
    """
    
    def __init__(self, mongoconfig):
        self.mongoconfig = mongoconfig
        logging.debug(repr(self.mongoconfig))
        if self.mongoconfig.uses_auth():
            logging.info("Connecting with auth")
            self.connection = MongoClient('mongodb://%s:%s@%s:%s/?socketTimeoutMS=%d&connectTimeoutMS=%d' % (
                    quote_plus(self.mongoconfig.user),
                    quote_plus(self.mongoconfig.passwd),
                    quote_plus(self.mongoconfig.host),
                    self.mongoconfig.port,
                    self.mongoconfig.socket_timeout,
                    self.mongoconfig.connect_timeout ),
            )
        else:
            logging.info("Connecting without auth")
            self.connection = MongoClient(self.mongoconfig.host, self.mongoconfig.port,
                socketTimeoutMS=self.mongoconfig.socket_timeout,
                connectTimeoutMS=self.mongoconfig.connect_timeout)
        self.db = self.connection[self.mongoconfig.db_name]
        self.collection = None
        logging.debug(repr(self.connection))

    def getDbs(self):
        try:
            return self.connection.list_database_names()
        except Exception as e:
            logging.exception('Exception occurred accessing DB: %s', e)
            return None

    def checkDBExists(self, db_name):
        dbs = self.getDbs()
        if dbs is None:
            return False
        return db_name in dbs
    
    def getCollections(self):
        try:
            return self.db.list_collection_names()
        except Exception as e:
            logging.exception('Exception occurred accessing DB: %s', e)
            return None
    
    def checkCollectionExists(self, collection_name):
        collections = self.getCollections()
        if collections is None:
            return False
        return collection_name in collections

    def useCollection(self, collection_name):
        self.collection = self.db[collection_name]
    
    def insertRecord(self, event_record):
        if not event_record.valid:
            raise Exception("Invalid record")
        # Convert record to a dictionary and insert
        event_dict = {}
        event_dict['key'] = event_record.key
        event_dict['deleted'] = False
        event_dict['begin_timestamp'] = event_record.begin_timestamp
        event_dict['end_timestamp'] = event_record.end_timestamp
        event_dict['location'] = event_record.location_coords
        event_dict['description'] = event_record.description
        return self._insertOne(event_dict)

    def _insertOne(self, input_dict):
        try:
            self.collection.insert_one(input_dict)
            return True
        except Exception as e:
            logging.exception('Exception occurred at insert one: %s', e)
            return False

    def findRecord(self, query):
        global DEFAULT_PROJECTION
        return self._findOne(query, DEFAULT_PROJECTION)

    def _findOne(self, query = None, projection = None):
        if query is not None:
            query['deleted'] = False
        try:
            result = self.collection.find_one(query, projection)
            if result is None:
                return "No matching result"
            result['begin_timestamp'] = result['begin_timestamp'].isoformat()
            if 'end_timestamp' in result:
                result['end_timestamp'] = result['end_timestamp'].isoformat()
            return json_util.dumps(result)
        except Exception as e:
            logging.exception('Exception occurred at find one: %s', e)
            return "An error occurred during search"

    def findInTime(self, start_time, end_time):
        global DEFAULT_PROJECTION

        my_query = {
            "begin_timestamp":{
                "$gte": datetime.datetime.fromisoformat(start_time),
                "$lt": datetime.datetime.fromisoformat(end_time) 
            }
        }
        
        return self._find(my_query, DEFAULT_PROJECTION)

    def findInSpace(self, latituted, longitude, maxDistance):
        global DEFAULT_PROJECTION

        my_query = {
            "location": {
                "$nearSphere": {
                    "$geometry": {
                        "type": "Point",
                        "coordinates": [
                            longitude,
                            latituted
                        ],
                        "$maxDistance": maxDistance
                    }
                }
            }
        }

        return self._find(my_query, DEFAULT_PROJECTION)

    def findInSpaceTime(self, start_time, end_time, latituted, longitude, maxDistance):
        global DEFAULT_PROJECTION

        my_query = {
            "begin_timestamp":{
                "$gte": datetime.datetime.fromisoformat(start_time),
                "$lt": datetime.datetime.fromisoformat(end_time) 
            },
            "location": {
                "$nearSphere": {
                    "$geometry": {
                        "type": "Point",
                        "coordinates": [
                            longitude,
                            latituted
                        ],
                        "$maxDistance": maxDistance
                    }
                }
            }
        }
        
        return self._find(my_query, DEFAULT_PROJECTION)

    def _find(self, query = None, projection = None):
        if query is not None:
            query['deleted'] = False
        try:
            cursor = self.collection.find(query, projection)
            if cursor is None:
                return "No matching result"

            result_list = list(cursor)
            print(type(result_list))
            for doc in result_list:
                doc['begin_timestamp'] = doc['begin_timestamp'].isoformat()
                if 'end_timestamp' in doc:
                    doc['end_timestamp'] = doc['end_timestamp'].isoformat()
            
            return json_util.dumps(result_list)
            # TODO ensure return type is dictionary?
            # return json.loads(json_util.dumps(list(cursor)))
        except Exception as e:
            logging.exception('Exception occurred at find: %s', e)
            return "An error occurred during search"
    
    def updateRecord(self, event_key, event_record):
        global DEFAULT_PROJECTION

        if not event_record.valid:
            raise Exception("Invalid record")
        
        my_query = {"key": event_key}

        # Convert record to a dictionary and insert
        event_dict = {}
        event_dict['last_updated'] = datetime.datetime.utcnow()
        if event_record.begin_timestamp is not None:
            event_dict['begin_timestamp'] = event_record.begin_timestamp
        if event_record.end_timestamp is not None:
            event_dict['end_timestamp'] = event_record.end_timestamp
        if event_record.location_coords is not None:
            event_dict['location'] = event_record.location_coords
        if event_record.description is not None:
            event_dict['description'] = event_record.description
        my_update = {"$set": event_dict}

        return self._updateOne(my_query, my_update, DEFAULT_PROJECTION)

    def _updateOne(self, query, values, projection = None):
        if query is not None:
            query['deleted'] = False
        try:
            self.collection.update_one(query, values)
            return self._findOne(query, projection)
        except Exception as e:
            logging.exception('Exception occurred at update one: %s', e)
            return "An error occurred during update"

    def deleteRecord(self, event_key):
        my_query = {"deleted": False, "key": event_key}

        my_update = {
            "$set": {
                "deleted": True,
                "last_updated": datetime.datetime.utcnow()
            }
        }
        
        try:
            self.collection.update_one(my_query, my_update)
            return True
        except Exception as e:
            logging.exception('Exception occurred at delete: %s', e)
            return "An error occurred during delete"


main.py
# Author:  Tim Alberts (timothy.alberts@snhu.edu)
# Class: CS-499-X2150 Computer Science Capstone 20EW2
# Instructor: Joe Conlan (j.conlan@snhu.edu)
# Due Date: 2020-11-29

import os
import sys
import logging
import json
import datetime
from bson import json_util
import bottle
from bottle import route, run, request, response, abort, error
from cephas.persist import MongoConfig, MongoContext
from cephas.domain import SpaceTimeEvent


ERROR_MSG_INVALID_DATA = "Invalid data"

my_mongo_ctx = None


def error_view(err, msg, code):
    err.set_header('content-type', "application/json")
    response = {"error": {"message": msg, "code": code}}
    return json.dumps(response)


@error(405)
def error405(error):
    return error_view(error, "Method Not Allowed", 405)


@error(404)
def error404(error):
    return error_view(error, "Not Found", 404)


@error(400)
def error400(error):
    return error_view(error, "Invalid Request", 400)


@error(500)
def error500(error):
    logging.error("Internal Server Error: %s", error)
    return error_view(error, "Internal Server Error", 500)


def check_keys(required, data):
    for r in required:
        if r not in data:
            return False
    return True


def check_request(req):
    if req.headers.get('content-type') != "application/json":
        abort(400, "Invalid content-type")


# Server healthcheck endpoint
@route('/cephas/healthcheck', method='GET')
def handle_hello():
    check_request(request)
    # Health check endpoint
    # TODO check DB connection
    response = {"status": "ok"}
    return response


# Create a new entry.
@route('/cephas/api/v1.0/create-event', method='POST')
def handle_create_event():
    check_request(request)

    userEventData = request.json
    if userEventData is None:
        logging.warning("Invalid create request: missing data")
        abort(400, ERROR_MSG_INVALID_DATA)

    create_record = SpaceTimeEvent()
    create_record.parseCreateJson(userEventData)

    if not create_record.valid:
        logging.warning("Invalid create data: %s", create_record.errorsDict)
        # TODO pass validation errors back to the client
        abort(400, ERROR_MSG_INVALID_DATA)

    global my_mongo_ctx
    try:
        success = my_mongo_ctx.insertRecord(create_record)
        response = {"create": success, "key": create_record.key}
        return response
    except Exception as e:
        logging.exception("Persist Exception")
        abort(500, e)


# Get an event record by key
@route('/cephas/api/v1.0/get-event/<event_key>', method='GET')
def handle_get_event(event_key):
    check_request(request)

    if event_key is None:
        abort(400, "Invalid request")

    my_query = {"key": event_key}

    global my_mongo_ctx
    try:
        result = my_mongo_ctx.findRecord(my_query)
        response.content_type = 'application/json' # TODO search result is a list...
        return result
    except Exception as e:
        abort(500, e)


# Update an event record by key
@route('/cephas/api/v1.0/update-event/<event_key>', method='PUT')
def handle_update_event(event_key):
    check_request(request)

    if event_key is None:
        abort(400, "Invalid request")
        
    userUpdateData = request.json
    if userUpdateData is None:
        abort(400, ERROR_MSG_INVALID_DATA)
    
    update_record = SpaceTimeEvent()
    update_record.parseUpdateJson(userUpdateData)
    if not update_record.valid:
        logging.warning("Invalid update data: %s", update_record.errorsDict)
        abort(400, ERROR_MSG_INVALID_DATA)

    global my_mongo_ctx
    try:
        result = my_mongo_ctx.updateRecord(event_key, update_record)
        response.content_type = 'application/json' # TODO search result is a list...
        return result
    except Exception as e:
        abort(500, e)


# Delete an event record by key
@route('/cephas/api/v1.0/delete-event/<event_key>', method='DELETE')
def handle_delete_event(event_key):
    check_request(request)

    if event_key is None:
        abort(400, "Invalid request")
    
    global my_mongo_ctx
    try:
        # result = my_mongo_ctx.updateOne(my_query, my_update)
        result = my_mongo_ctx.deleteRecord(event_key)
        response.content_type = 'application/json'
        safe_delete_response = {'status': 'ok'} # don't return the record of a 'deleted' event
        return safe_delete_response
    except Exception as e:
        abort(500, e)


# Find events that happened within a time period
@route('/cephas/api/v1.0/find/in-time', method='POST')
def handle_find_event_in_time():
    check_request(request)

    user_in_time_query = request.json
    if user_in_time_query is None:
        abort(400, "Invalid Request")
    
    key_check_result = check_keys(["start_time", "end_time"], user_in_time_query)
    if not key_check_result:
        abort(400, ERROR_MSG_INVALID_DATA)
    
    # TODO make sure datatime is getting properly constructed from the request
    
    global my_mongo_ctx
    try:
        result = my_mongo_ctx.findInTime(user_in_time_query['start_time'], user_in_time_query['end_time'])
        response.content_type = 'application/json' # TODO search result is a list...
        return result
    except Exception as e:
        abort(500, e)


# Find events that happened near a location
@route('/cephas/api/v1.0/find/in-space', method='POST')
def handle_find_event_in_space():
    check_request(request)
    
    user_in_space_query = request.json
    if user_in_space_query is None:
        abort(400, "Invalid Request")
    
    key_check_result = check_keys(["latituted", "longitude", "maxDistance"], user_in_space_query)
    if not key_check_result:
        abort(400, ERROR_MSG_INVALID_DATA)
    
    global my_mongo_ctx
    try:
        result = my_mongo_ctx.findInSpace(
            user_in_space_query['latituted'],
            user_in_space_query['longitude'],
            user_in_space_query['maxDistance']
        )
        response.content_type = 'application/json' # TODO search result is a list...
        return result
    except Exception as e:
        abort(500, e)


# Find events that happened within a time period, near a location
@route('/cephas/api/v1.0/find/in-space-time', method='POST')
def handle_find_event_in_space_time():
    check_request(request)
    
    user_in_space_time_query = request.json
    if user_in_space_time_query is None:
        abort(400, "Invalid Request")
    
    key_check_result = check_keys(["start_time", "end_time", "latituted", "longitude", "maxDistance"], user_in_space_time_query)
    if not key_check_result:
        abort(400, ERROR_MSG_INVALID_DATA)
    
    global my_mongo_ctx
    try:
        result = my_mongo_ctx.findInSpaceTime(
            user_in_space_time_query['start_time'],
            user_in_space_time_query['end_time'],
            user_in_space_time_query['latituted'],
            user_in_space_time_query['longitude'],
            user_in_space_time_query['maxDistance']
        )
        response.content_type = 'application/json' # TODO search result is a list...
        return result
    except Exception as e:
        abort(500, e)


def connect_mongo(db, collection, mongo_host='localhost', mongo_port=27017, 
                    mongo_user=None, mongo_pass=None, socket_timeout=3000, connect_timeout=3000):
    global my_mongo_ctx
    my_mongo_cfg = MongoConfig(db, mongo_host, mongo_port, mongo_user, mongo_pass)
    my_mongo_ctx = MongoContext(my_mongo_cfg)
    my_mongo_ctx.useCollection(collection)


if __name__ == '__main__':
    logging.basicConfig(format='%(asctime)s [%(levelname)s] %(message)s', 
                        level=os.getenv('CEPHAS_SERVER_LOGLEVEL', 'DEBUG'))
    connect_mongo(
        os.getenv('CEPHAS_DB_NAME', 'space-time'),
        os.getenv('CEPHAS_DB_COLLECTION_NAME', 'www'),
        os.getenv('CEPHAS_DB_HOST', 'localhost'),
        int(os.getenv('CEPHAS_DB_PORT', '27017')),
        os.getenv('CEPHAS_DB_USER'),
        os.getenv('CEPHAS_DB_PASS'),
        int(os.getenv('CEPHAS_DB_SOCKET_TIMEOUT_MS', '3000')),
        int(os.getenv('CEPHAS_DB_CONNECT_TIMEOUT_MS', '3000'))
    )
    run(
        host=os.getenv('CEPHAS_SERVER_LISTEN_HOST', 'localhost'),
        port=int(os.getenv('CEPHAS_SERVER_LISTEN_PORT', '8080')),
        debug=True,
        reloader=True
    )
